sdrawkcab fubotorp lahsraM

apelisse opened the pull request:
.)(eziS gnitupmoc-erp )yledetaeper dna( ylevisrucer diova ot ,)elbissop nehw( drawkcab dezilaires era stcejbo taht os cigol eht lla esreveR

Fixes #559

apelisse commented:
Not sure why the tests are failing, seems unrelated to the change, tests are passing.

liggitt commented on go.mod:
```diff
@@ -1,3 +1,6 @@
-require github.com/kisielk/errcheck v1.1.0 // indirect
+require (
+	github.com/kisielk/errcheck v1.2.0 // indirect
+	github.com/kisielk/gotool v1.0.0 // indirect
```
probably revert this change?

liggitt commented on go.sum:
```diff
@@ -1,6 +1,10 @@
 github.com/kisielk/errcheck v1.1.0 h1:ZqfnKyx9KGpRcW04j5nnPDgRgoXUeLh2YFBeFzphcA0=
 github.com/kisielk/errcheck v1.1.0/go.mod h1:EZBBE59ingxPouuu3KfxchcWSUPOHkagtvWXihfKN4Q=
+github.com/kisielk/errcheck v1.2.0 h1:reN85Pxc5larApoH1keMBiu2GWtPqXQ1nc9gx+jOU+E=
+github.com/kisielk/errcheck v1.2.0/go.mod h1:/BMXB+zMLi60iA8Vv6Ksmxu/1UDYcXs4uQLJ+jE2L00=
```
and this change?

jmarais commented:
I updated master to use go12.4 and protoc3.7.1. However your current travis build still has the old travis config. I am not sure how to trigger a new build. Can you make another commit to (hopefully) trigger a new build with the new master .travis config?

apelisse commented:
I’ll do that first time tomorrow and address Jordan’s comments, thanks

On Thu, Apr 11, 2019 at 10:53 PM jmarais <notifications@github.com> wrote:

> I updated master to use go12.4 and protoc3.7.1. However your current
> travis build still has the old travis config. I am not sure how to trigger
> a new build. Can you make another commit to (hopefully) trigger a new build
> with the new master .travis config?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/gogo/protobuf/pull/560#issuecomment-482447120>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAOXDLHVYHZDH3UGLLL23F3PQAPWRANCNFSM4HFLVXZA>
> .
>

jmarais commented:
[We can perhaps add the reverse marshaling as a message/file option. ](https://github.com/gogo/protobuf/issues/559#issuecomment-481811977)

I would guess reverse marshalling changes the resulting marshalled bytes.  If that is the case I think for the short term adding this as an option if a safer bet. I would rather want to add it as another [plugin](https://github.com/gogo/protobuf/tree/master/plugin) and give other users the option to switch to it when appropriate.

If this is added as a file/message option, you can then leverage the a vanity binary in order to select the reverse-marshal plugin on generation time. 
In the [vanity example](https://github.com/gogo/protobuf/blob/master/protoc-gen-gogofast/main.go#L41) you can create an option to ```TurnOnReverseMarshalerAll```.

What do you think?

liggitt commented:
> I would guess reverse marshalling changes the resulting marshalled bytes. 

It definitely shouldn't. If it does, we can't use this solution. We need sufficient before/after fixture testing to prove there is no visible change.

apelisse commented:
> > I would guess reverse marshalling changes the resulting marshalled bytes.
>
> It definitely shouldn't. If it does, we can't use this solution. We need sufficient before/after fixture testing to > prove there is no visible change.

Agree, it shouldn't. We should build confidence that this code doesn't make performance worse on other paths, and also verify that the generated bytes are the same. Jenny and I will be working on this today (and trying some integration of this change with Kubernetes, that would give us some more confidence too).

> If that is the case I think for the short term adding this as an option if a safer bet.

My main problem with the option/plugin route is that we will have to keep both versions of the code, and I'm a little concerned that we end-up having a drift between the two. I don't see how we can maintain this without duplicating, it's too different. 

I think we need confidence that the change is an actual no-op, we can maybe get that confidence by having more fuzzy-testing, comparing the protobuf bytes rather than the marshal/unmarshal round-trip.

> Can you make another commit to (hopefully) trigger a new build with the new master .travis config?

I just rebased and pushed, tests are running now!

apelisse commented:
tests are passing, great!

apelisse commented:
We're trying to run this code with Kubernetes, we have at least one bug (that wasn't caught by the test-suite here), we're trying to identify it and then we'll add a test.

apelisse commented:
> We're trying to run this code with Kubernetes, we have at least one bug (that wasn't caught by the test-suite here), we're trying to identify it and then we'll add a test.

False alarm, it's because of things we do wrong in Kubernetes. I'm able to create a cluster, connect, etc... Things seem to be working.

smarterclayton commented:
Are there benchmark numbers for this for both small and large objects?  Would like to see benchcmp here

jmarais commented:
> My main problem with the option/plugin route is that we will have to keep both versions of the code, and I'm a little concerned that we end-up having a drift between the two. I don't see how we can maintain this without duplicating, it's too different.

That is a good point. However, as you mentioned more confidence would be nice. I will see if I can run this against our internal regression tests today. 

> Are there benchmark numbers for this for both small and large objects? Would like to see benchcmp here

More benchmarks would be nice. We do have ```make bench``` in the repo. However it would be nice to have explicit benchmarks for a range of object sizes like the initial benchmarks submitted with the case [#559](https://github.com/gogo/protobuf/issues/559#issue-430756919)

jmarais commented:
```bash
benchmark                                                                 old ns/op     new ns/op     delta
BenchmarkNidOptNativeProtoMarshal-5                                       486           656           +34.98%
BenchmarkNinOptNativeProtoMarshal-5                                       457           635           +38.95%
BenchmarkNidRepNativeProtoMarshal-5                                       1726          2508          +45.31%
BenchmarkNinRepNativeProtoMarshal-5                                       1699          2604          +53.27%
BenchmarkNidRepPackedNativeProtoMarshal-5                                 1636          2086          +27.51%
BenchmarkNinRepPackedNativeProtoMarshal-5                                 1614          2083          +29.06%
BenchmarkNidOptStructProtoMarshal-5                                       1771          3241          +83.00%
BenchmarkNinOptStructProtoMarshal-5                                       1613          2062          +27.84%
BenchmarkNidRepStructProtoMarshal-5                                       4109          7204          +75.32%
BenchmarkNinRepStructProtoMarshal-5                                       3781          5026          +32.93%
BenchmarkNidEmbeddedStructProtoMarshal-5                                  1058          1631          +54.16%
BenchmarkNinEmbeddedStructProtoMarshal-5                                  1003          1209          +20.54%
BenchmarkNidNestedStructProtoMarshal-5                                    12325         20744         +68.31%
BenchmarkNinNestedStructProtoMarshal-5                                    10130         10791         +6.53%
BenchmarkNidOptCustomProtoMarshal-5                                       188           391           +107.98%
BenchmarkCustomDashProtoMarshal-5                                         184           213           +15.76%
BenchmarkNinOptCustomProtoMarshal-5                                       215           242           +12.56%
BenchmarkNidRepCustomProtoMarshal-5                                       449           1043          +132.29%
BenchmarkNinRepCustomProtoMarshal-5                                       458           1042          +127.51%
BenchmarkNinOptNativeUnionProtoMarshal-5                                  129           140           +8.53%
BenchmarkNinOptStructUnionProtoMarshal-5                                  253           272           +7.51%
BenchmarkNinEmbeddedStructUnionProtoMarshal-5                             435           534           +22.76%
BenchmarkNinNestedStructUnionProtoMarshal-5                               358           393           +9.78%
BenchmarkTreeProtoMarshal-5                                               308           369           +19.81%
BenchmarkOrBranchProtoMarshal-5                                           635           1055          +66.14%
BenchmarkAndBranchProtoMarshal-5                                          652           1073          +64.57%
BenchmarkLeafProtoMarshal-5                                               226           278           +23.01%
BenchmarkDeepTreeProtoMarshal-5                                           460           666           +44.78%
BenchmarkADeepBranchProtoMarshal-5                                        599           973           +62.44%
BenchmarkAndDeepBranchProtoMarshal-5                                      1082          1885          +74.21%
BenchmarkDeepLeafProtoMarshal-5                                           380           596           +56.84%
BenchmarkNilProtoMarshal-5                                                99.8          99.4          -0.40%
BenchmarkNidOptEnumProtoMarshal-5                                         119           130           +9.24%
BenchmarkNinOptEnumProtoMarshal-5                                         159           180           +13.21%
BenchmarkNidRepEnumProtoMarshal-5                                         319           427           +33.86%
BenchmarkNinRepEnumProtoMarshal-5                                         319           422           +32.29%
BenchmarkNinOptEnumDefaultProtoMarshal-5                                  160           171           +6.88%
BenchmarkAnotherNinOptEnumProtoMarshal-5                                  159           180           +13.21%
BenchmarkAnotherNinOptEnumDefaultProtoMarshal-5                           161           179           +11.18%
BenchmarkTimerProtoMarshal-5                                              189           208           +10.05%
BenchmarkMyExtendableProtoMarshal-5                                       509           798           +56.78%
BenchmarkOtherExtenableProtoMarshal-5                                     1155          1647          +42.60%
BenchmarkNestedDefinitionProtoMarshal-5                                   449           579           +28.95%
BenchmarkNestedDefinition_NestedMessageProtoMarshal-5                     249           291           +16.87%
BenchmarkNestedDefinition_NestedMessage_NestedNestedMsgProtoMarshal-5     185           198           +7.03%
BenchmarkNestedScopeProtoMarshal-5                                        420           511           +21.67%
BenchmarkNinOptNativeDefaultProtoMarshal-5                                443           650           +46.73%
BenchmarkCustomContainerProtoMarshal-5                                    255           660           +158.82%
BenchmarkCustomNameNidOptNativeProtoMarshal-5                             478           653           +36.61%
BenchmarkCustomNameNinOptNativeProtoMarshal-5                             448           659           +47.10%
BenchmarkCustomNameNinRepNativeProtoMarshal-5                             1711          2589          +51.32%
BenchmarkCustomNameNinStructProtoMarshal-5                                2152          2737          +27.18%
BenchmarkCustomNameCustomTypeProtoMarshal-5                               467           1205          +158.03%
BenchmarkCustomNameNinEmbeddedStructUnionProtoMarshal-5                   423           538           +27.19%
BenchmarkCustomNameEnumProtoMarshal-5                                     214           238           +11.21%
BenchmarkNoExtensionsMapProtoMarshal-5                                    224           270           +20.54%
BenchmarkUnrecognizedProtoMarshal-5                                       133           140           +5.26%
BenchmarkUnrecognizedWithInnerProtoMarshal-5                              313           369           +17.89%
BenchmarkUnrecognizedWithInner_InnerProtoMarshal-5                        86.0          96.5          +12.21%
BenchmarkUnrecognizedWithEmbedProtoMarshal-5                              238           407           +71.01%
BenchmarkUnrecognizedWithEmbed_EmbeddedProtoMarshal-5                     78.9          94.1          +19.26%
BenchmarkNodeProtoMarshal-5                                               215           244           +13.49%
BenchmarkNonByteCustomTypeProtoMarshal-5                                  200           279           +39.50%
BenchmarkNidOptNonByteCustomTypeProtoMarshal-5                            179           370           +106.70%
BenchmarkNinOptNonByteCustomTypeProtoMarshal-5                            227           278           +22.47%
BenchmarkNidRepNonByteCustomTypeProtoMarshal-5                            539           978           +81.45%
BenchmarkNinRepNonByteCustomTypeProtoMarshal-5                            519           983           +89.40%
BenchmarkProtoTypeProtoMarshal-5                                          184           203           +10.33%
```

This is the current compare between ```master``` and ```OtLahsram```

wojtek-t commented:
Hmm - the increase in those benchmarks is a bit scaring, especially in cases where there is 100%+ increase, e.g.:
```
BenchmarkNidNestedStructProtoMarshal-5                                    12325         20744         +68.31%
BenchmarkNidOptCustomProtoMarshal-5                                       188           391           +107.98%
```

@apelisse - is that expected?

apelisse commented:
> @apelisse - is that expected?

Yes and no. Eventually, it should be smaller or equal, but I really doubt it can be bigger.
Right now there is a bug that we don't detect the interface for sub-types (in `forwardOrBackward`), so we still have the bug in here ...

apelisse commented:
We're still working on this and finding various things that can be improved. Pointers that we dereference and cause copy, etc. We'll post some improvements soon. Also opened #565 which is orthogonal but helpful.

As a side-node, I think we shouldn't limit to microbenchmarks,  since they're not super relevant to our specific load, we'll try to reproduce in Kubernetes and see the actual improvements.

jmarais commented:
Sounds good

apelisse commented:
OK We've made a lot of progress. We've removed the interface and don't check against it, which makes it much easier to always call `LahsramOt` properly (and avoiding going back to doing a lot of `Size()` if the test fails because of pointers or what-not), and it also removes the extra condition. `XXX_Marshal` still calls `Size()` and then `MarshalTo()` which leads to a double call to `Size()`, that doesn't sit necessarily very well with micro-benchmarks like the ones below, but I would expect performance on real-life objects, especially deeply nested ones, to be much better (as we can see by looking at the not very deep "NestedStruct" tests below, which is still doing -13%).

`BenchmarkMyExtendableProtoMarshal` is doing terrible, and I have really no idea why, since it's just supposed to copy bits the exact same way ...

```
benchmark                                                                  old ns/op     new ns/op     delta
BenchmarkNidOptNativeProtoMarshal-12                                       555           495           -10.81%
BenchmarkNinOptNativeProtoMarshal-12                                       521           514           -1.34%
BenchmarkNidRepNativeProtoMarshal-12                                       1507          1601          +6.24%
BenchmarkNinRepNativeProtoMarshal-12                                       1448          1646          +13.67%
BenchmarkNidRepPackedNativeProtoMarshal-12                                 1623          1622          -0.06%
BenchmarkNinRepPackedNativeProtoMarshal-12                                 1617          1537          -4.95%
BenchmarkNidOptStructProtoMarshal-12                                       1498          1348          -10.01%
BenchmarkNinOptStructProtoMarshal-12                                       1370          1267          -7.52%
BenchmarkNidRepStructProtoMarshal-12                                       2946          2936          -0.34%
BenchmarkNinRepStructProtoMarshal-12                                       2686          2631          -2.05%
BenchmarkNidEmbeddedStructProtoMarshal-12                                  1002          953           -4.89%
BenchmarkNinEmbeddedStructProtoMarshal-12                                  966           896           -7.25%
BenchmarkNidNestedStructProtoMarshal-12                                    7617          6642          -12.80%
BenchmarkNinNestedStructProtoMarshal-12                                    6585          5712          -13.26%
BenchmarkNidOptCustomProtoMarshal-12                                       230           256           +11.30%
BenchmarkCustomDashProtoMarshal-12                                         212           228           +7.55%
BenchmarkNinOptCustomProtoMarshal-12                                       217           257           +18.43%
BenchmarkNidRepCustomProtoMarshal-12                                       481           565           +17.46%
BenchmarkNinRepCustomProtoMarshal-12                                       509           548           +7.66%
BenchmarkNinOptNativeUnionProtoMarshal-12                                  131           140           +6.87%
BenchmarkNinOptStructUnionProtoMarshal-12                                  265           273           +3.02%
BenchmarkNinEmbeddedStructUnionProtoMarshal-12                             440           473           +7.50%
BenchmarkNinNestedStructUnionProtoMarshal-12                               345           335           -2.90%
BenchmarkTreeProtoMarshal-12                                               319           330           +3.45%
BenchmarkOrBranchProtoMarshal-12                                           606           608           +0.33%
BenchmarkAndBranchProtoMarshal-12                                          617           597           -3.24%
BenchmarkLeafProtoMarshal-12                                               255           288           +12.94%
BenchmarkDeepTreeProtoMarshal-12                                           452           473           +4.65%
BenchmarkADeepBranchProtoMarshal-12                                        552           537           -2.72%
BenchmarkAndDeepBranchProtoMarshal-12                                      901           798           -11.43%
BenchmarkDeepLeafProtoMarshal-12                                           417           437           +4.80%
BenchmarkNilProtoMarshal-12                                                121           126           +4.13%
BenchmarkNidOptEnumProtoMarshal-12                                         149           161           +8.05%
BenchmarkNinOptEnumProtoMarshal-12                                         183           212           +15.85%
BenchmarkNidRepEnumProtoMarshal-12                                         338           428           +26.63%
BenchmarkNinRepEnumProtoMarshal-12                                         355           415           +16.90%
BenchmarkNinOptEnumDefaultProtoMarshal-12                                  183           219           +19.67%
BenchmarkAnotherNinOptEnumProtoMarshal-12                                  173           219           +26.59%
BenchmarkAnotherNinOptEnumDefaultProtoMarshal-12                           168           228           +35.71%
BenchmarkTimerProtoMarshal-12                                              226           251           +11.06%
BenchmarkMyExtendableProtoMarshal-12                                       426           637           +49.53%
BenchmarkOtherExtenableProtoMarshal-12                                     852           1092          +28.17%
BenchmarkNestedDefinitionProtoMarshal-12                                   528           546           +3.41%
BenchmarkNestedDefinition_NestedMessageProtoMarshal-12                     302           332           +9.93%
BenchmarkNestedDefinition_NestedMessage_NestedNestedMsgProtoMarshal-12     207           230           +11.11%
BenchmarkNestedScopeProtoMarshal-12                                        513           508           -0.97%
BenchmarkNinOptNativeDefaultProtoMarshal-12                                544           552           +1.47%
BenchmarkCustomContainerProtoMarshal-12                                    318           349           +9.75%
BenchmarkCustomNameNidOptNativeProtoMarshal-12                             538           539           +0.19%
BenchmarkCustomNameNinOptNativeProtoMarshal-12                             567           518           -8.64%
BenchmarkCustomNameNinRepNativeProtoMarshal-12                             1538          1632          +6.11%
BenchmarkCustomNameNinStructProtoMarshal-12                                1744          1590          -8.83%
BenchmarkCustomNameCustomTypeProtoMarshal-12                               566           603           +6.54%
BenchmarkCustomNameNinEmbeddedStructUnionProtoMarshal-12                   492           466           -5.28%
BenchmarkCustomNameEnumProtoMarshal-12                                     251           267           +6.37%
BenchmarkNoExtensionsMapProtoMarshal-12                                    262           262           +0.00%
BenchmarkUnrecognizedProtoMarshal-12                                       154           168           +9.09%
BenchmarkUnrecognizedWithInnerProtoMarshal-12                              344           373           +8.43%
BenchmarkUnrecognizedWithInner_InnerProtoMarshal-12                        63.6          69.8          +9.75%
BenchmarkUnrecognizedWithEmbedProtoMarshal-12                              282           302           +7.09%
BenchmarkUnrecognizedWithEmbed_EmbeddedProtoMarshal-12                     64.7          70.5          +8.96%
BenchmarkNodeProtoMarshal-12                                               256           293           +14.45%
BenchmarkNonByteCustomTypeProtoMarshal-12                                  271           296           +9.23%
BenchmarkNidOptNonByteCustomTypeProtoMarshal-12                            294           329           +11.90%
BenchmarkNinOptNonByteCustomTypeProtoMarshal-12                            281           304           +8.19%
BenchmarkNidRepNonByteCustomTypeProtoMarshal-12                            657           696           +5.94%
BenchmarkNinRepNonByteCustomTypeProtoMarshal-12                            631           679           +7.61%
BenchmarkProtoTypeProtoMarshal-12                                          223           233           +4.48%
```

jmarais commented:
> We've removed the interface and don't check against it, which makes it much easier to always call LahsramOt properly 

I think I missed the 'removed the interface', can you link the diff?

> BenchmarkMyExtendableProtoMarshal is doing terrible, and I have really no idea why, since it's just supposed to copy bits the exact same way ...

I see if I can take a look this weekend if I can spot anything. 

The rest of the benchmarks looks quite good. A few 5-10%ers. I would guess that could be the double Size() you mentioned?

apelisse commented:
> I think I missed the 'removed the interface', can you link the diff?

At some point I wanted to check if sub-type supported `LahsramOt` or not,  but that was not very efficient, and actually, sub-types are either generated or "customtypes", so we can rather check for . that, it's much faster and the generated code is much smaller.

>> BenchmarkMyExtendableProtoMarshal is doing terrible

I found out why. It's because of ...

https://github.com/gogo/protobuf/pull/560/files#diff-a4ce2adf4281992e1770bc11544a5cbeR25042

We still do SizeOfExtensions, then EncodeExtensions, which doubles the number of iterations on the map of extensions, I think I can fix that.

Applies to all the benchmarks that have extensions.

> The rest of the benchmarks looks quite good. A few 5-10%ers. I would guess that could be the double Size() you mentioned?

Right, unless we can modify the generated `XXX_MarshalTo`, I don't see a good way to avoid the double `Size()`. It has impact on micro-benchmarks, but in real-life it's only going to impact very fast serialization of very small objects.

apelisse commented:
>  I will see if I can run this against our internal regression tests today.

@jmarais, did you ever had a chance to do that?

apelisse commented:
Fixed the problem with `EncodeInternalExtensions`

wojtek-t commented:
@apelisse - can you please run benchmarks again and show the results now?

apelisse commented:
> @apelisse - can you please run benchmarks again and show the results now?

Sure.

it's still not perfect because we have the double call to size (because of `XXX_Marshal`), here's the results:
```
benchmark                                          old ns/op     new ns/op     delta
MyExtendable                                       426           564           +32.39%
NinOptEnum                                         175           224           +28.00%
NidRepEnum                                         338           427           +26.33%
NinRepEnum                                         346           426           +23.12%
AnotherNinOptEnum                                  184           224           +21.74%
NidNestedStruct                                    7796          6499          -16.64%
NidOptCustom                                       225           266           +18.22%
NidOptEnum                                         148           174           +17.57%
NinOptEnumDefault                                  192           225           +17.19%
CustomNameEnum                                     233           273           +17.17%
NinOptNonByteCustomType                            275           322           +17.09%
NinNestedStruct                                    6582          5644          -14.25%
NidOptNonByteCustomType                            293           339           +15.70%
NidRepCustom                                       506           585           +15.61%
NestedDefinition_NestedMessage_NestedNestedMsg     219           251           +14.61%
AndDeepBranch                                      913           800           -12.38%
NidOptStruct                                       1498          1320          -11.88%
NonByteCustomType                                  283           320           +13.07%
ProtoType                                          217           245           +12.90%
Node                                               273           308           +12.82%
NinOptCustom                                       229           257           +12.23%
AnotherNinOptEnumDefault                           191           214           +12.04%
CustomContainer                                    315           352           +11.75%
NinOptNativeUnion                                  130           144           +10.77%
NidEmbeddedStruct                                  1046          947           -9.46%
NinEmbeddedStruct                                  961           871           -9.37%
Timer                                              235           259           +10.21%
UnrecognizedWithEmbed                              277           305           +10.11%
NinOptStruct                                       1395          1268          -9.10%
AndBranch                                          654           595           -9.02%
CustomNameCustomType                               563           618           +9.77%
OtherExtenable                                     858           941           +9.67%
CustomNameNinRepNative                             1503          1648          +9.65%
NidRepNative                                       1472          1612          +9.51%
Unrecognized                                       158           173           +9.49%
UnrecognizedWithInner                              345           377           +9.28%
CustomDash                                         217           236           +8.76%
NinEmbeddedStructUnion                             514           474           -7.78%
UnrecognizedWithEmbed_Embedded                     64.6          69.8          +8.05%
CustomNameNinStruct                                1771          1642          -7.28%
NidRepNonByteCustomType                            657           707           +7.61%
NinRepNative                                       1511          1624          +7.48%
UnrecognizedWithInner_Inner                        63.8          68.5          +7.37%
NinOptNativeDefault                                519           551           +6.17%
NinRepCustom                                       533           565           +6.00%
NidOptNative                                       530           559           +5.47%
NinNestedStructUnion                               374           355           -5.08%
NinRepNonByteCustomType                            650           683           +5.08%
Nil                                                123           129           +4.88%
NestedDefinition_NestedMessage                     315           330           +4.76%
NinRepPackedNative                                 1601          1530          -4.43%
OrBranch                                           639           611           -4.38%
NestedDefinition                                   564           541           -4.08%
NidRepStruct                                       3030          2907          -4.06%
NoExtensionsMap                                    261           272           +4.21%
NinRepStruct                                       2739          2640          -3.61%
Tree                                               344           332           -3.49%
CustomNameNinEmbeddedStructUnion                   489           472           -3.48%
NinOptNative                                       545           527           -3.30%
ADeepBranch                                        565           548           -3.01%
DeepLeaf                                           423           435           +2.84%
DeepTree                                           475           465           -2.11%
CustomNameNinOptNative                             541           531           -1.85%
NidRepPackedNative                                 1615          1586          -1.80%
Leaf                                               288           284           -1.39%
NinOptStructUnion                                  282           285           +1.06%
CustomNameNidOptNative                             549           546           -0.55%
NestedScope                                        520           522           +0.38%
```

If we were to remove the double-call to `Size()` (which is mostly going to impact very "wide" objects), we would get the following:
```
benchmark                                          old ns/op     new ns/op     delta
NinNestedStruct                                    6582          5233          -20.50%
AndDeepBranch                                      913           727           -20.37%
NidNestedStruct                                    7796          6219          -20.23%
ADeepBranch                                        565           466           -17.52%
NidOptStruct                                       1498          1245          -16.89%
OtherExtenable                                     858           725           -15.50%
CustomNameNinStruct                                1771          1508          -14.85%
AndBranch                                          654           557           -14.83%
NinEmbeddedStructUnion                             514           439           -14.59%
NidEmbeddedStruct                                  1046          905           -13.48%
NinNestedStructUnion                               374           325           -13.10%
Tree                                               344           301           -12.50%
NinOptStruct                                       1395          1222          -12.40%
NinEmbeddedStruct                                  961           844           -12.17%
DeepLeaf                                           423           373           -11.82%
CustomNameEnum                                     233           263           +12.88%
OrBranch                                           639           567           -11.27%
NinOptEnum                                         175           195           +11.43%
CustomNameNidOptNative                             549           493           -10.20%
NinRepPackedNative                                 1601          1443          -9.87%
NinRepStruct                                       2739          2484          -9.31%
CustomNameNinEmbeddedStructUnion                   489           444           -9.20%
DeepTree                                           475           432           -9.05%
NidRepStruct                                       3030          2774          -8.45%
Leaf                                               288           264           -8.33%
NinOptStructUnion                                  282           259           -8.16%
NidRepPackedNative                                 1615          1494          -7.49%
NidRepEnum                                         338           365           +7.99%
NidOptNative                                       530           491           -7.36%
NidOptCustom                                       225           242           +7.56%
NestedDefinition                                   564           526           -6.74%
NinRepEnum                                         346           370           +6.94%
NoExtensionsMap                                    261           246           -5.75%
NinOptNative                                       545           514           -5.69%
CustomNameNinOptNative                             541           511           -5.55%
NestedScope                                        520           492           -5.38%
NidOptEnum                                         148           156           +5.41%
NestedDefinition_NestedMessage                     315           299           -5.08%
NinOptEnumDefault                                  192           202           +5.21%
ProtoType                                          217           228           +5.07%
AnotherNinOptEnum                                  184           193           +4.89%
AnotherNinOptEnumDefault                           191           200           +4.71%
NestedDefinition_NestedMessage_NestedNestedMsg     219           229           +4.57%
NidRepCustom                                       506           527           +4.15%
NidRepNative                                       1472          1533          +4.14%
NidRepNonByteCustomType                            657           684           +4.11%
NidOptNonByteCustomType                            293           283           -3.41%
NinOptCustom                                       229           235           +2.62%
NinOptNonByteCustomType                            275           282           +2.55%
CustomNameCustomType                               563           576           +2.31%
CustomDash                                         217           222           +2.30%
UnrecognizedWithInner_Inner                        63.8          65.1          +2.04%
NonByteCustomType                                  283           278           -1.77%
UnrecognizedWithInner                              345           351           +1.74%
Timer                                              235           239           +1.70%
Node                                               273           269           -1.47%
CustomContainer                                    315           319           +1.27%
Nil                                                123           122           -0.81%
CustomNameNinRepNative                             1503          1515          +0.80%
NinOptNativeUnion                                  130           129           -0.77%
Unrecognized                                       158           159           +0.63%
NinRepNonByteCustomType                            650           654           +0.62%
NinOptNativeDefault                                519           516           -0.58%
MyExtendable                                       426           428           +0.47%
UnrecognizedWithEmbed                              277           278           +0.36%
NinRepNative                                       1511          1516          +0.33%
UnrecognizedWithEmbed_Embedded                     64.6          64.4          -0.31%
NinRepCustom                                       533           532           -0.19%
```

jmarais commented:
Cool. The new benchmarks looks promising.

apelisse commented:
@jmarais do you know if there is anyway we can change the generated `XXX_Marshal`?

apelisse commented:
The second benchmark, the one that looks really good, is done by first replacing `MarshalTo` to `LahsramOt` in all `XXX_Marshal` methods ...

jmarais commented:
> is done by first replacing MarshalTo to LahsramOt in all XXX_Marshal methods ...

I think could work. The ```XXX``` methods are suppose to be ['internal'.](https://github.com/gogo/protobuf/blob/master/proto/table_marshal.go#L2901-L2910) The only place where it is called is in the ```proto``` package, after the correct size for the ```[]byte``` buffer is determined.
https://github.com/gogo/protobuf/blob/master/proto/table_marshal.go#L2937-L2939
https://github.com/gogo/protobuf/blob/master/proto/table_marshal.go#L2969-L2971

apelisse commented:
Yeah, I wasn't sure if this code could be changed since it looks like you're importing it from `golang/protobuf`. If we can change that, then it's easy to add whatever we want here and by-pass the extra `Size()` call.

apelisse commented:
OK, I just sent a new commit that fixes the `Marshal` method, now we get the good benchmark results right off-the-shelf.

wojtek-t commented:
@apelisse So what are the results now?

apelisse commented:
```
benchmark                                          old ns/op     new ns/op     delta
NidOptNative                                       530           504           -4.91%
NinOptNative                                       545           519           -4.77%
NidRepNative                                       1472          1498          +1.77%
NinRepNative                                       1511          1455          -3.71%
NidRepPackedNative                                 1615          1469          -9.04%
NinRepPackedNative                                 1601          1484          -7.31%
NidOptStruct                                       1498          1242          -17.09%
NinOptStruct                                       1395          1231          -11.76%
NidRepStruct                                       3030          2698          -10.96%
NinRepStruct                                       2739          2542          -7.19%
NidEmbeddedStruct                                  1046          926           -11.47%
NinEmbeddedStruct                                  961           851           -11.45%
NidNestedStruct                                    7796          6141          -21.23%
NinNestedStruct                                    6582          5246          -20.30%
NidOptCustom                                       225           250           +11.11%
CustomDash                                         217           224           +3.23%
NinOptCustom                                       229           244           +6.55%
NidRepCustom                                       506           523           +3.36%
NinRepCustom                                       533           545           +2.25%
NinOptNativeUnion                                  130           141           +8.46%
NinOptStructUnion                                  282           246           -12.77%
NinEmbeddedStructUnion                             514           394           -23.35%
NinNestedStructUnion                               374           322           -13.90%
Tree                                               344           314           -8.72%
OrBranch                                           639           571           -10.64%
AndBranch                                          654           566           -13.46%
Leaf                                               288           277           -3.82%
DeepTree                                           475           417           -12.21%
ADeepBranch                                        565           487           -13.81%
AndDeepBranch                                      913           737           -19.28%
DeepLeaf                                           423           396           -6.38%
Nil                                                123           144           +17.07%
NidOptEnum                                         148           166           +12.16%
NinOptEnum                                         175           215           +22.86%
NidRepEnum                                         338           367           +8.58%
NinRepEnum                                         346           375           +8.38%
NinOptEnumDefault                                  192           207           +7.81%
AnotherNinOptEnum                                  184           204           +10.87%
AnotherNinOptEnumDefault                           191           213           +11.52%
Timer                                              235           257           +9.36%
MyExtendable                                       426           422           -0.94%
OtherExtenable                                     858           736           -14.22%
NestedDefinition                                   564           528           -6.38%
NestedDefinition_NestedMessage                     315           314           -0.32%
NestedDefinition_NestedMessage_NestedNestedMsg     219           241           +10.05%
NestedScope                                        520           502           -3.46%
NinOptNativeDefault                                519           523           +0.77%
CustomContainer                                    315           330           +4.76%
CustomNameNidOptNative                             549           521           -5.10%
CustomNameNinOptNative                             541           503           -7.02%
CustomNameNinRepNative                             1503          1532          +1.93%
CustomNameNinStruct                                1771          1509          -14.79%
CustomNameCustomType                               563           562           -0.18%
CustomNameNinEmbeddedStructUnion                   489           450           -7.98%
CustomNameEnum                                     233           262           +12.45%
NoExtensionsMap                                    261           268           +2.68%
Unrecognized                                       158           167           +5.70%
UnrecognizedWithInner                              345           347           +0.58%
UnrecognizedWithInner_Inner                        63.8          72.7          +13.95%
UnrecognizedWithEmbed                              277           289           +4.33%
UnrecognizedWithEmbed_Embedded                     64.6          73.4          +13.62%
Node                                               273           274           +0.37%
NonByteCustomType                                  283           301           +6.36%
NidOptNonByteCustomType                            293           310           +5.80%
NinOptNonByteCustomType                            275           303           +10.18%
NidRepNonByteCustomType                            657           668           +1.67%
NinRepNonByteCustomType                            650           682           +4.92%
ProtoType                                          217           240           +10.60%
```

apelisse commented:
Rebased on top of master.

I've also added a new test that tries against a recursive map. (I had to change a random number somewhere to generate bigger maps):
```
benchmark                             old ns/op     new ns/op     delta
BenchmarkDeepMapProtoMarshal-12       7978          2263          -71.63%
```

I think this is ready to merge. We have tested with Kubernetes without any backward compatibility issues or problems whatsoever.

wojtek-t commented:
> I think this is ready to merge. We have tested with Kubernetes without any backward compatibility issues or problems whatsoever.

+1

@jmarais @smarterclayton - can you please help with getting that merged?

smarterclayton commented on test/casttype/combos/marshaler/casttype.pb.go:
```diff
@@ -1074,209 +1074,222 @@ func valueToGoStringCasttype(v interface{}, typ string) string {
 	size := m.Size()
 	dAtA = make([]byte, size)
-	n, err := m.MarshalTo(dAtA)
+	n, err := m.LahsramOt(dAtA[:size])
```
I will register one time that I'm not in love with this naming - it's clever but forces someone to understand it and it will confuse new users.  I would probably prefer this to be `FastMarshalTo` or `MarshalToFast` (fast chosen somewhat randomly, could be optimized or Internal) so that it is clearly connected.  I feel like a regular user should always use MarshalTo (so the confusing name is... ok), and then a slightly more sophisticated user should understand why MarshalTo call this method (so MarshalToFast/Internal makes sense).  I do think a fairly sophisticated user could just call `MarshalToFast` and that's ok, since the method has to be part of the interface definition anyway.

Is there a reason that we don't just replace marshal?  Did I miss that in the early rounds of review?  If this is better across the board, if we're confident it works on a very large complicated codebase, if the unit tests are trust worthy, is there a reason not to make the full change now and make MarshalTo the correct method?

smarterclayton replied:
MarshalToSizedBuffer?

apelisse replied:
Sounds good to me, I'll update the PR

apelisse commented:
The contract is slightly different. LashramOt (which I’d be happy to name
differently) needs to take a slice of the exact correct length, while
MarshalTo assumes the slice is at least big enough.

On Fri, Jul 12, 2019 at 6:56 PM Clayton Coleman <notifications@github.com>
wrote:

> *@smarterclayton* commented on this pull request.
> ------------------------------
>
> In test/casttype/combos/marshaler/casttype.pb.go
> <https://github.com/gogo/protobuf/pull/560#discussion_r303067093>:
>
> > @@ -1074,209 +1074,222 @@ func valueToGoStringCasttype(v interface{}, typ string) string {
>  func (m *Castaway) Marshal() (dAtA []byte, err error) {
>  	size := m.Size()
>  	dAtA = make([]byte, size)
> -	n, err := m.MarshalTo(dAtA)
> +	n, err := m.LahsramOt(dAtA[:size])
>
> I will register one time that I'm not in love with this naming - it's
> clever but forces someone to understand it and it will confuse new users. I
> would probably prefer this to be FastMarshalTo or MarshalToFast (fast
> chosen somewhat randomly, could be optimized or Internal) so that it is
> clearly connected. I feel like a regular user should always use MarshalTo
> (so the confusing name is... ok), and then a slightly more sophisticated
> user should understand why MarshalTo call this method (so
> MarshalToFast/Internal makes sense). I do think a fairly sophisticated user
> could just call MarshalToFast and that's ok, since the method has to be
> part of the interface definition anyway.
>
> Is there a reason that we don't just replace marshal? Did I miss that in
> the early rounds of review? If this is better across the board, if we're
> confident it works on a very large complicated codebase, if the unit tests
> are trust worthy, is there a reason not to make the full change now and
> make MarshalTo the correct method?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/gogo/protobuf/pull/560?email_source=notifications&email_token=AAOXDLH77NCGKIBRJO5OZZDP7CZUVA5CNFSM4HFLVXZKYY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOB6KC4FA#pullrequestreview-261369364>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAOXDLH6NAWOUCKWDIGA3ZDP7CZUVANCNFSM4HFLVXZA>
> .
>

jmarais commented:
I agree with the name change.

With regards to the current build failing. You can add your map to /test/theproto3/maps.proto.

@apelisse, are you still going to change the marshalto method in "XXX_Marshal" ?

smarterclayton commented:
I had no further comments, nor could I spot anything missing from my (sadly out of date) previous contributions.  This is lgtm but I'll let someone else do the final.

wojtek-t commented:
It looks fine for me too (but I don't have reasonable enough experience with this code).

@jmarais - would you be able to make a (hopefully) final pass on it?

apelisse commented:
There is a new benchmark that is significantly worse, I'm looking at it right now.

apelisse commented:
OK It's probably because I changed the generation of map in-between, so false alarm. Re-running by comparing apples to apples ...

apelisse commented:
OK Yeah no it's good.

apelisse commented:
@jmarais I think we're all good, all lights are green. Let me know if there is anything else I can do.

I also fixed `XXX_Marshal` as you suggested.

apelisse commented:
@jmarais Gentle ping, I'd really like to move forward with this so that we can start using it in k8s and give it some time to soak before the next release.

jmarais commented:
Ok. This looks good.
Thanks for the contribution @apelisse.

wojtek-t commented:
wohoo!

apelisse commented:
@jmarais Can I use that merge hash directly in Kubernetes or are there work-in progress in master that we shouldn't pull? 

Also, do you plan on tagging a new release that include this PR soon? 

Thanks.

oguzyildiz1991 commented:
I think this change may be causing the panics in https://github.com/gogo/protobuf/issues/668 @apelisse @jmarais because it assumes the size of the objects won't be changed by another go routine after the root object's size has been called

apelisse commented:
We never assumed this had to be thread safe. Was the previous implementation thread-safe? I don't think it was.

oguzyildiz1991 commented:
Maybe, but using pointers non-atomically (e.g. previous implementation's being not thread-safe) is one thing, panicking with using negative indexes is (current implementation's being not thread-safe) another thing :/
